File: services/taskService.js
import { TaskRepository } from '../repositories/taskRepository.js';
import { Task } from '../models/Task.js';

export class TaskService {
  constructor() {
    this.taskRepository = new TaskRepository();
  }

  // Create a new task
  async createTask(taskData) {
    try {
      // Additional business logic can be added here
      if (!taskData.title || taskData.title.trim() === '') {
        throw new Error('Task title is required');
      }

      const createdTask = await this.taskRepository.create(taskData);
      return new Task(createdTask);
    } catch (error) {
      throw error;
    }
  }

  // Get task by ID
  async getTaskById(id) {
    try {
      if (!id) {
        throw new Error('Task ID is required');
      }

      const task = await this.taskRepository.getById(id);
      if (!task) {
        throw new Error('Task not found');
      }

      return new Task(task);
    } catch (error) {
      throw error;
    }
  }

  // Get all tasks with filtering and pagination
  async getAllTasks(filters = {}) {
    try {
      // Set default values and validate pagination
      const page = Math.max(1, parseInt(filters.page) || 1);
      const pageSize = Math.min(100, Math.max(1, parseInt(filters.pageSize) || 10));

      const queryFilters = {
        ...filters,
        page,
        pageSize
      };

      const result = await this.taskRepository.getAll(queryFilters);
      
      return {
        tasks: result.tasks.map(task => new Task(task)),
        total: result.total,
        page,
        pageSize,
        totalPages: Math.ceil(result.total / pageSize)
      };
    } catch (error) {
      throw error;
    }
  }

  // Update task by ID
  async updateTask(id, updateData) {
    try {
      if (!id) {
        throw new Error('Task ID is required');
      }

      // Check if task exists
      const exists = await this.taskRepository.exists(id);
      if (!exists) {
        throw new Error('Task not found');
      }

      // Additional business logic validation
      if (updateData.title !== undefined && (!updateData.title || updateData.title.trim() === '')) {
        throw new Error('Task title cannot be empty');
      }

      const updatedTask = await this.taskRepository.update(id, updateData);
      if (!updatedTask) {
        throw new Error('Task not found');
      }

      return new Task(updatedTask);
    } catch (error) {
      throw error;
    }
  }

  // Delete task by ID
  async deleteTask(id) {
    try {
      if (!id) {
        throw new Error('Task ID is required');
      }

      // Check if task exists
      const exists = await this.taskRepository.exists(id);
      if (!exists) {
        throw new Error('Task not found');
      }

      await this.taskRepository.delete(id);
      return true;
    } catch (error) {
      throw error;
    }
  }

  // Get task statistics (additional business logic)
  async getTaskStats() {
    try {
      const allTasks = await this.taskRepository.getAll({ pageSize: 1000 });
      const tasks = allTasks.tasks;

      const stats = {
        total: tasks.length,
        pending: tasks.filter(t => t.status === 'pending').length,
        in_progress: tasks.filter(t => t.status === 'in_progress').length,
        completed: tasks.filter(t => t.status === 'completed').length,
        high_priority: tasks.filter(t => t.priority === 'high').length,
        medium_priority: tasks.filter(t => t.priority === 'medium').length,
        low_priority: tasks.filter(t => t.priority === 'low').length,
        overdue: tasks.filter(t => t.due_date && new Date(t.due_date) < new Date()).length
      };

      return stats;
    } catch (error) {
      throw error;
    }
  }
}

File: services/task_service.go
package services

import (
	"fmt"
	"task-api/models"
	"task-api/repositories"

	"github.com/google/uuid"
)

type TaskService struct {
	repo *repositories.TaskRepository
}

// NewTaskService creates a new task service
func NewTaskService() *TaskService {
	return &TaskService{
		repo: repositories.NewTaskRepository(),
	}
}

// CreateTask creates a new task
func (s *TaskService) CreateTask(req *models.CreateTaskRequest) (*models.Task, error) {
	// Additional business logic validation can be added here
	if req.Title == "" {
		return nil, fmt.Errorf("task title is required")
	}

	return s.repo.Create(req)
}

// GetTaskByID retrieves a task by its ID
func (s *TaskService) GetTaskByID(id uuid.UUID) (*models.Task, error) {
	return s.repo.GetByID(id)
}

// GetAllTasks retrieves all tasks with optional filtering and pagination
func (s *TaskService) GetAllTasks(status, priority string, page, pageSize int) ([]models.Task, int, error) {
	// Set default pagination values
	if page < 1 {
		page = 1
	}
	if pageSize < 1 || pageSize > 100 {
		pageSize = 10
	}

	offset := (page - 1) * pageSize
	return s.repo.GetAll(status, priority, pageSize, offset)
}

// UpdateTask updates an existing task
func (s *TaskService) UpdateTask(id uuid.UUID, req *models.UpdateTaskRequest) (*models.Task, error) {
	// Check if task exists
	_, err := s.repo.GetByID(id)
	if err != nil {
		return nil, err
	}

	// Additional business logic validation can be added here
	if req.Title != nil && *req.Title == "" {
		return nil, fmt.Errorf("task title cannot be empty")
	}

	return s.repo.Update(id, req)
}

// DeleteTask deletes a task by its ID
func (s *TaskService) DeleteTask(id uuid.UUID) error {
	// Check if task exists
	_, err := s.repo.GetByID(id)
	if err != nil {
		return err
	}

	return s.repo.Delete(id)
}